Duplicati/GUI/Duplicati.GUI.TrayIcon/Program.cs

```diff
@@ -133,73 +137,256 @@ public static void RealMain(string[] _args)
+
+                do
                 {
-                    if (hosted == null)
+                    try
                     {
-                        url = DEFAULT_HOSTURL;
+                        using (Connection = new HttpServerConnection(serverURL, password, saltedpassword))
+                        {
+                            using (var tk = RunTrayIcon(toolkit))
+                            {
+                                if (hosted != null && Server.Program.Instance != null)
+                                    Server.Program.Instance.SecondInstanceDetected +=
+                                        new Server.SingleInstance.SecondInstanceDelegate(
+                                            x => { tk.ShowUrlInWindow(serverURL.ToString()); });
+
+                                // TODO: If we change to hosted browser this should be a callback
+                                if (openui)
+                                {
+                                    try
+                                    {
+                                        tk.ShowUrlInWindow(Connection.StatusWindowURL);
+
+                                        Duplicati.Server.Program.IsFirstRun = false;
+                                        Duplicati.Server.Program.ServerPortChanged = false;
+                                    }
+                                    catch
+                                    {
+                                    }
+                                }
+
+                                // If the server shuts down, shut down the tray-icon as well
+                                Action shutdownEvent = () =>
+                                {
+                                    tk.InvokeExit();
+                                };
+
+                                if (hosted != null)
+                                    hosted.InstanceShutdown += shutdownEvent;
+
+                                tk.Init(_args);
+
+                                // Make sure that the server shutdown does not access the tray-icon,
+                                // as it would be disposed by now
+                                if (hosted != null)
+                                    hosted.InstanceShutdown -= shutdownEvent;
+                            }
+                        }
                     }
-                    else
+                    catch (WebException ex)
                     {
-                        int port = Duplicati.Server.Program.ServerPort;
-                        url = "http://127.0.0.1:" + port;
+                        System.Diagnostics.Trace.WriteLine("Request error: " + ex.Message);
+                        Console.WriteLine("Request error: " + ex.Message);
+
+                        reSpawn++;
                     }
-                }
+                } while (reSpawn < 3);
+            }
+        }
+
+        private static Server.Database.Connection GetDatabaseConnection(string[] args)

```

Comment:
Why not make `GetDatabaseConnection` public in `Duplicati.Server` and then call it there?